cli/tests/pcluster/aws/dummy_aws_api.py

```diff
@@ -337,8 +342,8 @@ def describe_secret(self, secret_arn):
         }
 
 
-def mock_aws_api(mocker, mock_instance_type_info=True):
+def mock_aws_api(mocker, mock_instance_type_info=True, non_happy=False):
     """Mock AWS Api."""
-    mocker.patch("pcluster.aws.aws_api.AWSApi.instance", return_value=_DummyAWSApi())
+    mocker.patch("pcluster.aws.aws_api.AWSApi.instance", return_value=_DummyAWSApi(non_happy))
```

Comment:
You added the new parameter `non_happy` to `_DummyAwsApi` to inject the unhappy behaviour into the FSx client. The problems with this approach are:
1. the `non_happy` parameter should inject the unhappy behaviour into the FSx client only, but your are setting at the overall API level, so from the outside one could expect that the unhappy behaviour would affect also other APIs; In case this would be the correct approach (but it is not), the name of the parameter should have been more specific to the failing API, e.g. `non_happy_fsx`. 
2. in this unit test we want to inject an unhappy behaviour specifically into the FSx API, without any other impact on other APIs, so I would expect the `mocker.patch` method to patch the specific FSx function, e.g. `pcluster.aws.fsx.FSxClient.describe_volumes` (or whatever other API call you want to mock);

Reply:
I agree with your first point, however, I disagree about it's impact to other API calls as the argument is optional and therefore does not affect any existing code. Furthermore, this approach enables faster implementation of similar non-happy tests in the future. As for your second point, I agree that would be ideal, however that does not work as expected. Using the exact example you gave fails because the actual `FSxClient` from `fsx` is not used due to being replaced in the mocking logic. Furthermore, a similar approach fails i.e. attempting to modify the exact mocked `describe_volumes` function because the testing code, as seen in the validator approach's implementation, requires an actual function from source code to mock. So, passing the argument for non-happy flows through the mocking architecture is the simplest, correct, solution that I've found.

Reply:
I understand the challenge you're facing.
We can overcome this blocker with the approach I proposed in the above comment:

```
def your_unit_test(mocker):
    dummy_api = _DummyAWSApi()
    dummy_api._fsx.make_describe_volumes_fail_with_error(True, error=UnauthorizedException)
    mocker.patch("pcluster.aws.aws_api.AWSApi.instance", return_value=dummy_api)

   ... your unit tests logic ...
```

Reply:
Ok